firebase-perf/src/main/java/com/google/firebase/perf/application/AppStateMonitor.java

```diff
@@ -149,28 +150,64 @@ public void incrementTsnsCount(int value) {
     tsnsCount.addAndGet(value);
   }
 
-  @Override
-  public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
+  // Starts tracking the frame metrics for an activity.
+  public void startFrameMonitoring(Activity activity) {
     if (isScreenTraceSupported() && configResolver.isPerformanceMonitoringEnabled()) {
+      FrameMetricsRecorder recorder = new FrameMetricsRecorder(activity);
+      activityToRecorderMap.put(activity, recorder);
       if (activity instanceof FragmentActivity) {
+        FragmentStateMonitor fragmentStateMonitor =
+            new FragmentStateMonitor(clock, transportManager, this, recorder);
+        activityToFragmentStateMonitorMap.put(activity, fragmentStateMonitor);
         FragmentActivity fragmentActivity = (FragmentActivity) activity;
         fragmentActivity
             .getSupportFragmentManager()
-            .registerFragmentLifecycleCallbacks(
-                new FragmentStateMonitor(clock, transportManager, this, frameMetricsAggregator),
-                true);
+            .registerFragmentLifecycleCallbacks(fragmentStateMonitor, true);
       }
     }
   }
 
   @Override
-  public void onActivityDestroyed(Activity activity) {}
+  public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
+    startFrameMonitoring(activity);
+  }
 
   @Override
-  public synchronized void onActivityStarted(Activity activity) {}
+  public void onActivityDestroyed(Activity activity) {
+    // Dereference FrameMetricsRecorder from the map because it holds an Activity reference
+    activityToRecorderMap.remove(activity);
+    // Dereference FragmentStateMonitor because it holds a FrameMetricsRecorder reference
+    if (activityToFragmentStateMonitorMap.containsKey(activity)) {
+      FragmentActivity fragmentActivity = (FragmentActivity) activity;
+      fragmentActivity
+          .getSupportFragmentManager()
+          .unregisterFragmentLifecycleCallbacks(activityToFragmentStateMonitorMap.remove(activity));
+    }
+  }
+
+  @Override
+  public synchronized void onActivityStarted(Activity activity) {
+    if (isScreenTraceSupported() && configResolver.isPerformanceMonitoringEnabled()) {
+      if (activityToRecorderMap.get(activity) == null) {
```

Comment:
This is more directly expressed as `activityToRecorderMap.containsKey(activity)`

Reply:
Got it. Done!